File: common/src/main/java/juuxel/adorn/recipe/ItemBrewingRecipe.java
package juuxel.adorn.recipe;

import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import juuxel.adorn.block.AdornBlocks;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.ItemStackTemplate;
import net.minecraft.world.item.crafting.Ingredient;
import net.minecraft.world.item.crafting.PlacementInfo;
import net.minecraft.world.item.crafting.RecipeSerializer;
import net.minecraft.world.item.crafting.display.RecipeDisplay;
import net.minecraft.world.item.crafting.display.SlotDisplay;
import net.minecraft.world.level.Level;

import java.util.List;
import java.util.Optional;

import static juuxel.adorn.block.entity.BrewerBlockEntity.*;

public record ItemBrewingRecipe(
    CommonInfo commonInfo,
    Ingredient input,
    Ingredient firstIngredient,
    Optional<Ingredient> secondIngredient,
    ItemStackTemplate result
) implements BrewingRecipe {
    public static final MapCodec<ItemBrewingRecipe> CODEC = RecordCodecBuilder.mapCodec(builder -> builder.group(
        CommonInfo.MAP_CODEC.forGetter(ItemBrewingRecipe::commonInfo),
        Ingredient.CODEC.fieldOf("input").forGetter(ItemBrewingRecipe::input),
        Ingredient.CODEC.fieldOf("first_ingredient").forGetter(ItemBrewingRecipe::firstIngredient),
        Ingredient.CODEC.optionalFieldOf("second_ingredient").forGetter(ItemBrewingRecipe::secondIngredient),
        ItemStackTemplate.CODEC.fieldOf("result").forGetter(ItemBrewingRecipe::result)
    ).apply(builder, ItemBrewingRecipe::new));

    private static final StreamCodec<RegistryFriendlyByteBuf, ItemBrewingRecipe> STREAM_CODEC = StreamCodec.composite(
        CommonInfo.STREAM_CODEC,
        ItemBrewingRecipe::commonInfo,
        Ingredient.CONTENTS_STREAM_CODEC,
        ItemBrewingRecipe::input,
        Ingredient.CONTENTS_STREAM_CODEC,
        ItemBrewingRecipe::firstIngredient,
        Ingredient.OPTIONAL_CONTENTS_STREAM_CODEC,
        ItemBrewingRecipe::secondIngredient,
        ItemStackTemplate.STREAM_CODEC,
        ItemBrewingRecipe::result,
        ItemBrewingRecipe::new
    );

    public static final RecipeSerializer<ItemBrewingRecipe> SERIALIZER = new RecipeSerializer<>(CODEC, STREAM_CODEC);

    @Override
    public boolean matches(BrewerInput input, Level world) {
        var ingredientsMatch = (input.matches(LEFT_INGREDIENT_SLOT, firstIngredient) && input.matches(RIGHT_INGREDIENT_SLOT, secondIngredient)) ||
            (input.matches(RIGHT_INGREDIENT_SLOT, firstIngredient) && input.matches(LEFT_INGREDIENT_SLOT, secondIngredient));
        return ingredientsMatch && input.matches(INPUT_SLOT, this.input);
    }

    @Override
    public ItemStack assemble(BrewerInput input) {
        return result.create();
    }

    @Override
    public boolean showNotification() {
        return commonInfo.showNotification();
    }

    @Override
    public String group() {
        return "";
    }

    @Override
    public RecipeSerializer<ItemBrewingRecipe> getSerializer() {
        return SERIALIZER;
    }

    @Override
    public List<RecipeDisplay> display() {
        return List.of(
            new BrewingRecipeDisplay(
                input.display(),
                firstIngredient.display(),
                secondIngredient.map(Ingredient::display).orElse(SlotDisplay.Empty.INSTANCE),
                SlotDisplay.Empty.INSTANCE,
                new SlotDisplay.ItemStackSlotDisplay(result),
                new SlotDisplay.ItemSlotDisplay(AdornBlocks.BREWER.get().asItem())
            )
        );
    }

    @Override
    public PlacementInfo placementInfo() {
        return PlacementInfo.createFromOptionals(List.of(Optional.of(input), Optional.of(firstIngredient), secondIngredient));
    }
}

File: common/src/main/kotlin/juuxel/adorn/recipe/ItemBrewingRecipe.kt
package juuxel.adorn.recipe

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import juuxel.adorn.block.entity.BrewerBlockEntity.Companion.LEFT_INGREDIENT_SLOT
import juuxel.adorn.block.entity.BrewerBlockEntity.Companion.RIGHT_INGREDIENT_SLOT
import net.minecraft.item.ItemStack
import net.minecraft.network.PacketByteBuf
import net.minecraft.recipe.Ingredient
import net.minecraft.recipe.RecipeSerializer
import net.minecraft.registry.DynamicRegistryManager
import net.minecraft.world.World

class ItemBrewingRecipe(
    val firstIngredient: Ingredient,
    val secondIngredient: Ingredient,
    val result: ItemStack
) : BrewingRecipe {
    override fun matches(inventory: BrewerInventory, world: World): Boolean {
        fun matches(index: Int, ingredient: Ingredient) =
            ingredient.test(inventory.getStack(index))

        return (matches(LEFT_INGREDIENT_SLOT, firstIngredient) && matches(RIGHT_INGREDIENT_SLOT, secondIngredient)) ||
            (matches(RIGHT_INGREDIENT_SLOT, firstIngredient) && matches(LEFT_INGREDIENT_SLOT, secondIngredient))
    }

    override fun craft(inventory: BrewerInventory, registryManager: DynamicRegistryManager): ItemStack = result.copy()
    override fun fits(width: Int, height: Int): Boolean = true
    override fun getResult(registryManager: DynamicRegistryManager): ItemStack = result
    override fun getSerializer(): RecipeSerializer<*> = AdornRecipes.BREWING_SERIALIZER

    companion object {
        val CODEC: Codec<ItemBrewingRecipe> = RecordCodecBuilder.create { builder ->
            builder.group(
                Ingredient.DISALLOW_EMPTY_CODEC.fieldOf("first_ingredient").forGetter { it.firstIngredient },
                Ingredient.ALLOW_EMPTY_CODEC.optionalFieldOf("second_ingredient", Ingredient.empty()).forGetter { it.secondIngredient },
                ItemStack.RECIPE_RESULT_CODEC.fieldOf("result").forGetter { it.result }
            ).apply(builder, ::ItemBrewingRecipe)
        }
    }

    class Serializer : RecipeSerializer<ItemBrewingRecipe> {
        override fun codec(): Codec<ItemBrewingRecipe> = CODEC

        override fun read(buf: PacketByteBuf): ItemBrewingRecipe {
            val first = Ingredient.fromPacket(buf)
            val second = Ingredient.fromPacket(buf)
            val output = buf.readItemStack()
            return ItemBrewingRecipe(first, second, output)
        }

        override fun write(buf: PacketByteBuf, recipe: ItemBrewingRecipe) {
            recipe.firstIngredient.write(buf)
            recipe.secondIngredient.write(buf)
            buf.writeItemStack(recipe.result)
        }
    }
}
